### ISSUE TITLE
tsc compiler does not notice missing return item inside try catch block for async function

### ISSUE THREAD

**Issue by Orillion360** [Microsoft/TypeScript#14081]
**TypeScript Version:**  2.1.1 / nightly (2.2.0-dev.201xxxxx)

**Code**

```ts
// A *self-contained* demonstration of the problem follows...
public async DoSomethingAsync(): Promise<number> {
    try {
        await DoSomething();
    } catch (error) {
        return 1;
    }
}
```

**Expected behavior:**
The TSC compiler should throw an error because no number is returned inside the try {} block, only in the catch {} block

**Actual behavior:**
It does not throw an error and the Promise resolves to undefined when no error occurs at runtime

**Comment by mhegazy** [comment]
A function in JS at run time returns `undefined` implicitly if it does not have an explicit return statement. 

Using `--strictNullChecks` you should get an error that the function possibly returns `undefined` but the type annotation says it is a `number`.

without `strictNullChecks` both `undefined` and `null` are part pf the domain of all types.

Using `--noImplicitReturns` you should get an error for `Not all code paths returning a value`.
